### ISSUE TITLE
Cascading updates when using usePagination

### ISSUE THREAD

**Issue by kizu** [tannerlinsley/react-table#1775]
I've noticed that in some cases the table is rendered twice after some events like sorting or filtering, after some debugging it seems that this happens only when `usePagination` is present.

Example with sorting: https://codesandbox.io/s/tannerlinsleyreact-table-sorting-t47hd
Example with filters: https://codesandbox.io/s/bold-bose-1mx7l

There is a console.log, when the `usePagination` is present, it is called twice on sort/filtering, when there is no pagination, only once.

In the perf dev tools the second render is visible and seems to be caused by the cascading update: 
![image](https://user-images.githubusercontent.com/177485/70792353-4ef37000-1d99-11ea-80d7-d87c6d64bc75.png)

This is probably a bug somewhere in the pagination code? For larger tables I imagine this can make some actions be less responsive than it could be.

**Comment by tannerlinsley** [comment]
This happens when the auto reset effect is triggered. It’s by design and should be a relatively cheap render.

Tanner Linsley

**Comment by kizu** [comment]
For me it is not that cheap for larger tables (same basic table, 100 rows — https://codesandbox.io/s/tannerlinsleyreact-table-sorting-80bee):
![image](https://user-images.githubusercontent.com/177485/70830427-73c70200-1df0-11ea-91f6-a6e6e907430e.png)

Basically, I see this second render taking 50-100% of the original time, so it is x1.5-2.0 increase. It is good to know that disabling auto-reset removes this second render, but I wonder if it is still possible to fit everything into one, hmm.

Looking at the profiler, yes, the useTable hook itself in this case is very small (30ms before, <1 after, but most of the time goes into all the other stuff happening — like Reconciliation etc.):

![image](https://user-images.githubusercontent.com/177485/70830682-141d2680-1df1-11ea-90d1-2f9ae3077265.png)

**Comment by tannerlinsley** [comment]
Are you measuring in production by chance?

Tanner Linsley

**Comment by kizu** [comment]
This is on codesandbox, thus it is in development, in a production build things are a bit better, but there is still overhead.

Here is it with `autoResetPage: false`:
![image](https://user-images.githubusercontent.com/177485/70832209-d0c4b700-1df4-11ea-874d-43468da6c5c5.png)

Here it is with two renders, note how there is a bunch of stuff after the first onCommitFiberRoot:

![image](https://user-images.githubusercontent.com/177485/70832244-ecc85880-1df4-11ea-8389-9978d973266a.png)

Basically, when there is a second render, for this table with 100 rows there is another ~26ms of work, in that case this is close to 1/4–1/3 of the initial render. Not that drastic, but I imagine for larger roots and more complex components inside the cells this could be bigger just from all the extra reconciliation needed.

**Comment by tannerlinsley** [comment]
Yeah, the discussion around this has come up a lot in v7 but never really yielded a better approach thus far. Up to this point, I haven't been able to find a way to effectively update the internal table state *during* a render, instead of using a layout effect, while also keeping a relatively stable reducer model.

I would love any feedback on how to do this better, but for the "auto reset" functionality to work fluidly, I haven't been able to find a better way than placing it in a layout effect.

**Comment by kizu** [comment]
I'm not sure I would be able to look at the internals of all this, but maybe one day when I won't be busy with everything else :D Anyway, knowing that there is a way to opt-out of the auto-reset is already great, I'll see if this would be a real problem, or if the current state would be enough.

**Comment by tannerlinsley** [comment]
What's interesting is that if you opt-out, but need to achieve the same thing, most would just for the effect approach again in their component code. 🤷‍♂

**Comment by tannerlinsley** [comment]
Either way, feel free to keep commenting with more info or close the issue. Your call.

**Comment by kizu** [comment]
I think I'll close for now, thank you!

**Comment by kizu** [comment]
@tannerlinsley One small question: currently when the `autoResetPage: true`, the second render happens even when pageIndex is `0`. If I understand correctly, the goal of `autoResetPage` is to reset the pageIndex to its initial value when there are changes. Is this correct?

If so, then I think we can make a small optimisation and do not reset when we're already at the initial page? This would mean that there would be 2 renders if we were, say, at the second page, as we would need to go to the first one anyway, but if we were already at the first page and we get new data/sort/filter, then we don't need to return to a different pageIndex, and so we don't need to do this second render?

**Comment by tannerlinsley** [comment]
Indeed. A small check on the page index would be good here. Likewise for all of the other auto reset flags too

Tanner Linsley
